### PR TITLE
Fix Test_AccessController

### PR DESCRIPTION
**Pull request by pdbain-ibm**
Add special cases in custom classloaders, via a common superclass, to load test
harness classes.

Add a new security policy which allows the test harness to read and write the
results directory while the security manager is active.

Improve test progress reporting.

Delete obsolete comments, extract redundant strings to a constant.

Fixes issue #858.

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

### CODE REVIEW

**Comment by pdbain-ibm** [comment]
Other changes expected for Test_AccessControlContext.

**Comment by pdbain-ibm** [comment]
Copyright failure is in the Java security policy file.

File: test/Java8andUp/playlist.xml [hunk 1]
```diff
@@ -922,6 +922,69 @@
 		</subsets>
 	</test>
 
+	<test>
+		<featureIds>
+			<featureId>Java access control</featureId>
+		</featureIds>
+		<testCaseName>JCL_TEST_Java-Security_SE90</testCaseName>
```
Comment (smlambert) [inline test/Java8andUp/playlist.xml]: The final make target that gets generated takes testCaseName + Subset + NumberOfVariation (count starting from 0).  If you call the test this, the target name will be JCL_TEST_Java-Security_SE90_SE90_0  is that what you intend?  Same comment applies to testCaseName J9VMInternals_Test_SE80 below... J9VMInternals_Test_SE80_SE80_0 and J9VMInternals_Test_SE80_SE80_1,

Comment (pdbain-ibm) [reply test/Java8andUp/playlist.xml]: Shelley

In re your comment:

> the target name will be JCL_TEST_Java-Security_SE90_SE90_0

Could you please clarify?

Embedding the Java version appears to be not unusual in the playlist, cf. analogous existing entry <testCaseName>JCL_Test_SE90</testCaseName> (from which I extracted the new test cases), et al.

Looking at a recent internal build, I see:

make -f /bluebird/builds/bld_377668/jvmtest/test/SE90/TestConfig/wrapper.mk _JCL_Test_SE90_0 TEST_ROOT=...

and the generated makefile SE90/Java8andUp/autoGen.mk contains:

JCL_Test_SE90_0: TEST_RESROOT=$(JVM_TEST_ROOT)$(D)Java8andUp
JCL_Test_SE90_0: JVM_OPTIONS=$(RESERVED_OPTIONS) $(EXTRA_OPTIONS)
JCL_Test_SE90_0: TEST_GROUP=level.sanity
JCL_Test_SE90_0:
    @$(MKTREE) $(REPORTDIR);
    @echo "" | tee -a $(Q)$(TESTOUTPUT)$(D)TestTargetResult$(Q);
    @echo "===============================================" | tee -a $(Q)$(TESTOUTPUT)$(D)TestTargetResult$(Q);
    @echo "Running test $@ ..." | tee -a $(Q)$(TESTOUTPUT)$(D)TestTargetResult$(Q);
...



All that said, this is still WIP (and likely has other problems).
I'm trying to merge the two cases into one, so this may be moot.

@llxia @TianyuZuo FYI.

Comment (llxia) [reply test/Java8andUp/playlist.xml]: Just to clarify, we do not need to add SE90 or SE80 in test name. However, the test name has to be unique. One case I have seen people use SE80 and SE90 is that if they have one test that uses different commands for SE80 and SE90, people use testXXX_SE80 and testXXX_SE90 as test name.  In your case, since the test will only run in SE90, there is no need to have SE90 in the test name.

Comment (pdbain-ibm) [reply test/Java8andUp/playlist.xml]: Correction: this test runs in both Java 8 and Java 9. As it happens, I believe I can use the same command line for both, so the point is now moot.  Personal build in progress.

Comment (pdbain-ibm) [reply test/Java8andUp/playlist.xml]: ... or not:
  java.lang.AssertionError: unexpected: java.lang.reflect.InaccessibleObjectException: Unable to make field java.security.ProtectionDomain[] java.security.AccessControlContext.context accessible: module java.base does not "opens java.security" to unnamed module @6579a4e

Java 8 passes fine.

Looks like I need separate test cases with "--add-opens" for Java 9 and above.

Comment (pdbain-ibm) [reply test/Java8andUp/playlist.xml]: @smlambert @llxia please review.  
As note above, copyright failure is in the Java security policy file.

 Thanks.

File: test/Java8andUp/playlist.xml [hunk 2]
```diff
@@ -921,6 +919,84 @@
...
 		</subsets>
 	</test>
+<!--
+	Java 8 specific test case.
+	Java 9 and above require VM options not available in Java 8.
+-->
+	<test>
+		<featureIds>
+			<featureId>Java access control</featureId>
+		</featureIds>
```
Comment (llxia) [inline test/Java8andUp/playlist.xml]: Please remove <featureIds>.
We are in the process of removing <featureIds> in all playlist.

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessControlContext.java [hunk 3]
```diff
@@ -1,7 +1,7 @@
 package org.openj9.test.java.security;
 
 /*******************************************************************************
- * Copyright (c) 1998, 2017 IBM Corp. and others
+ * Copyright (c) 1998, 2018 IBM Corp. and others
```
Comment (llxia) [inline test/Java8andUp/src/org/openj9/test/java/security/Test_AccessControlContext.java]: Did we modify this file? Why we need to update the copyright?

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java [hunk 4]
```diff
@@ -494,17 +493,6 @@ public PermissionCollection getPermissions(CodeSource cs) {
 				Object ret = AccessController.doPrivileged(new PrivilegedAction<Object>() {
 					public Object run() {
 						try {
-							/* This tests a use scenario with following stacktrace from a standalone testcase
```
Comment (pshipton) [inline test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: Why is this removed?

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java [hunk 5]
```diff
@@ -517,28 +505,6 @@ public Object run() {
 				if (null == ret) {
 					Assert.fail("FAIL: TEST FAILED");
 				}
-				/*
```
Comment (pshipton) [inline test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: Why is this removed?

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java (hunk 4 above)
Comment (pdbain-ibm) [reply test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: I took out the stack traces because they clutter the code and don't add much value.
They also may not be accurate still.

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java (hunk 5 above)
Comment (pdbain-ibm) [reply test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: Ibid.

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java (hunk 4 above)
Comment (pshipton) [reply test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: They add value when understanding what the tests are doing, and when having to debug a failure. Please leave them in.

Comment (pdbain-ibm) [reply test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: Done.

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java [hunk 6]
```diff
@@ -353,17 +352,18 @@ public void test_doPrivilegedFrameStackWalking() {
...
 			String result = (String)java.security.AccessController.doPrivileged(pa);
 			Assert.fail("FAIL: security exception NOT thrown, get system property " + PROP_JAVA_HOME + " : " + result);
 		} catch (java.security.AccessControlException e) {
-			logger.debug("PASS: security exception expected.");
+			logger.debug(PASSED_SECURITY_EXCEPTION_EXPECTED);
 		}
 
-		// loop 20 time here such that the reflection object are optimized and bytecodes are created in a new class
-		// (after calling invoke() about 15 times), which introduces a slightly different reflection frame
+		/*
```
Comment (pshipton) [inline test/Java8andUp/src/org/openj9/test/java/security/Test_AccessController.java]: Just FYI there is nothing wrong with using // comments in Java code

**Comment by pdbain-ibm** [comment]
From the C/C++ coding standards (emphasis mine):

   Comments that span more than one line should use the Java style as
   follows:
   /* This is a long comment
       * that spans many lines
       * and this is how it
       * should look.
       */

(style quibble: should the "/*' not be on its own line?)

Also, from the "J9 Team Process" page:

   Java Coding Standards
   The coding standards for Java generally follow the C conventions with a
   few exceptions like braces to be more consistent with Java style.

---------------------------------------------------------------------
Peter Bain
Software Designer, IBM Runtime Technologies
IBM Ottawa Software Lab
3755 Riverside Drive, Ottawa, Ontario, K1V 1B8
613-356-5452
pdbain@ca.ibm.com

File: test/Java8andUp/src/org/openj9/test/java/security/Test_AccessControlContext.java (hunk 3 above)
Comment (pdbain-ibm) [reply test/Java8andUp/src/org/openj9/test/java/security/Test_AccessControlContext.java]: That was a mistake.  Change reverted.

File: test/Java8andUp/playlist.xml (hunk 2 above)
Comment (pdbain-ibm) [reply test/Java8andUp/playlist.xml]: Done.

**Comment by pdbain-ibm** [comment]
Do the reviewers have any further comments?
Thank you

**Comment by pshipton** [comment]
13:09:04 The following files were modified and have incorrect copyrights
[Pipeline] echo
13:09:04 test/Java8andUp/testacc.policy

File: test/Java8andUp/playlist.xml [hunk 7]
```diff
@@ -830,6 +828,78 @@
...
+<!--
+	Java 8 specific test case.
+	Java 9 and above require VM options not available in Java 8.
+-->
+	<test>
+		<testCaseName>JCL_TEST_Java-Security_SE80</testCaseName>
+		<variations>
+			<variation>NoOptions</variation>
+		</variations>
+		<command>$(MKDIR) -p $(REPORTDIR); \
```
Comment (llxia) [inline test/Java8andUp/playlist.xml]: We can use $(MKTREE) instead of $(MKDIR). Same for the test below.

**Comment by pdbain-ibm** [comment]
In re @pshipton 's comment: none of the other policy files have copyrights.  Is that deliberate?

**Comment by pshipton** [comment]
An oversight, not adding copyrights to the other policy files. Somebody could volunteer to fix them all, or they can be fixed as needed in order to pass the copyright check.

File: test/Java8andUp/testacc.policy [hunk 8]
```diff
@@ -0,0 +1,26 @@
+//  Copyright (c) 2016, 2018 IBM Corp. and others
```
Comment (pshipton) [inline test/Java8andUp/testacc.policy]: The start year should not be 2016 for a new file.

Comment (pdbain-ibm) [reply test/Java8andUp/testacc.policy]: Fixed.

**Comment by pdbain-ibm** [comment]
@smlambert ready for your review.

File: test/Java8andUp/testacc.policy [hunk 9]
```diff
@@ -0,0 +1,26 @@
+//  Copyright (c) 2018, 2018 IBM Corp. and others
```
Comment (smlambert) [inline test/Java8andUp/testacc.policy]: You are really taking Peter Shipton's comment about its okay to use // for comments seriously here for this 20 line block of commented code.  Out of curiosity, are policy files limited to using that style of comment or can /* */ also be used?

File: test/Java8andUp/playlist.xml [hunk 10]
```diff
@@ -830,6 +828,78 @@
 			<subset>SE80</subset>
 		</subsets>
 	</test>
+<!--
```
Comment (smlambert) [inline test/Java8andUp/playlist.xml]: Thought we are currently limited to not support comments in the playlist files (blocked by and awaiting an infra issue to have XML::Parser installed on test machines).  Good if its not causing issues, though not sure the comments adds any information of value, as its pretty clear by the test name and the list of options that they are different.  I would likely get rid of the comments from the playlist.

**Comment by smlambert** [comment]
Jenkins test sanity plinux

**Comment by smlambert** [comment]
Please update the branch to resolve the conflict (latest_exclude_SE90.txt).

**Comment by pdbain-ibm** [comment]
Done.
